### PR TITLE
Numerics docs page

### PR DESCRIPTION
**Pull request by bandish-shah**
* Numerics documentation for Composer 0.4.1 release

### CODE REVIEW

File: docs/source/trainer/numerics.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+For training, the most commonly used low precision formats are IEEE `fp16` and the Brain Floating Point `bf16` format.  While both formats utilize the same number of bits in memory, `bf16` offers a greater numerical range than `fp16` at the expense of being able to represent fewer decimal places, resulting in reduced sensitivity.
+
+Please see the following references for additioanl details on the common number formats:
+
+- `fp32`: [Single-precision Floating-point](https://en.wikipedia.org/wiki/Single-precision_floating-point_format)
+- `fp16`: [Half-precision Floating-point](https://en.wikipedia.org/wiki/Half-precision_floating-point_format)
+- `bf16`: [Brain Floating Point](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format)
```
Comment (bandish-shah) [inline docs/source/trainer/numerics.md:40]:
```suggestion
- `bf16`: [Brain Floating-point](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format)
```

File: docs/source/trainer/numerics.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+When discussing number formats, the precision generally refers to the number of total bits used to represent decimal digits (e.g., `fp32` represents numbers using 32 bits of computer memory).  Higher precision formats can represent a greater range of numbers and more decimal places compared to lower precision formats.  However, since lower precision formats utilize fewer bits, they occupy a smaller memory footprint.  Additionally, depending on the accelerator, lower precision formats can typically unlock more compute resulting in greater overall throughput.  Therefore it is often advantageous to use half-precision formats to accelerate training.
+
+For training, the most commonly used low precision formats are IEEE `fp16` and the Brain Floating Point `bf16` format.  While both formats utilize the same number of bits in memory, `bf16` offers a greater numerical range than `fp16` at the expense of being able to represent fewer decimal places, resulting in reduced sensitivity.
+
+Please see the following references for additioanl details on the common number formats:
+
+- `fp32`: [Single-precision Floating-point](https://en.wikipedia.org/wiki/Single-precision_floating-point_format)
```
Comment (ravi-mosaicml) [inline docs/source/trainer/numerics.md:38]: Nit: can you make these links inline in the above paragraph?

File: docs/source/trainer/numerics.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+- `fp16`: [Half-precision Floating-point](https://en.wikipedia.org/wiki/Half-precision_floating-point_format)
+- `bf16`: [Brain Floating-point](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format)
+
+### Mixed Precision
+
+Using half-precision number formats can boost model throughput, though not without issue.  The reduced representable range and sensitivity of these formats can cause several problems during training such as:
```
Comment (ravi-mosaicml) [inline docs/source/trainer/numerics.md:44]: Can you link to the pytorch docs that specify which ops are done in which precision when using mixed precision.

File: docs/source/trainer/numerics.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+For training, the most commonly used low precision formats are IEEE `fp16` and the Brain Floating Point `bf16` format.  While both formats utilize the same number of bits in memory, `bf16` offers a greater numerical range than `fp16` at the expense of being able to represent fewer decimal places, resulting in reduced sensitivity.
+
+Please see the following references for additioanl details on the common number formats:
+
+- `fp32`: [Single-precision Floating-point](https://en.wikipedia.org/wiki/Single-precision_floating-point_format)
+- `fp16`: [Half-precision Floating-point](https://en.wikipedia.org/wiki/Half-precision_floating-point_format)
+- `bf16`: [Brain Floating-point](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format)
+
+### Mixed Precision
```
Comment (ravi-mosaicml) [inline docs/source/trainer/numerics.md:42]: I'm wondering if it would make sense to merge the AMP section below with this one....main reasoning is that AMP is how someone would interact with composer.

```suggestion
### Automatic Mixed Precision (AMP)
```

File: docs/source/trainer/numerics.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+Using half-precision number formats can boost model throughput, though not without issue.  The reduced representable range and sensitivity of these formats can cause several problems during training such as:
+
+- Gradient underflow: Very small values can get zeroed out
+- Incorrect weight updates: Small gradient values can lead to no weight updates
+- Activation/Loss overflow: Very large values can roll over to representations of [Infinity](https://en.wikipedia.org/wiki/Infinity#Computing)(`Inf`)or [Not a Number](https://en.wikipedia.org/wiki/NaN) (`NaN`)
+
+The result is reduced model accuracy.
+
+The solution is to perform mixed precision training, where both single and half-precision formats are utilized strategically.  Mixed precision training is performed as follows:
```
Comment (ravi-mosaicml) [inline docs/source/trainer/numerics.md:52]:
```suggestion
The solution is to perform mixed precision training, where both single (32 bit) and half-precision (16 bit) formats are utilized strategically.  Mixed precision training is performed as follows:
```

File: docs/source/trainer/numerics.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+1. Compute the forward pass and loss in half-precision, *except* for computations that can cause activations/loss overflow, which are performed in single-precision (e.g., BatchNorm)
+2. Compute the backpropagation pass in half-precision
+3. To increase numerical sensitivity, convert the model to single-precision and perform the optimizer step, enabling the weight update to be done more precisely
+4. Convert the model back to half-precision
```
Comment (ravi-mosaicml) [inline docs/source/trainer/numerics.md:57]:
```suggestion
4. Convert the model back to half-precision.
```

File: docs/source/trainer/numerics.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+- Incorrect weight updates: Small gradient values can lead to no weight updates
+- Activation/Loss overflow: Very large values can roll over to representations of [Infinity](https://en.wikipedia.org/wiki/Infinity#Computing)(`Inf`)or [Not a Number](https://en.wikipedia.org/wiki/NaN) (`NaN`)
+
+The result is reduced model accuracy.
+
+The solution is to perform mixed precision training, where both single and half-precision formats are utilized strategically.  Mixed precision training is performed as follows:
+
+1. Compute the forward pass and loss in half-precision, *except* for computations that can cause activations/loss overflow, which are performed in single-precision (e.g., BatchNorm)
+2. Compute the backpropagation pass in half-precision
+3. To increase numerical sensitivity, convert the model to single-precision and perform the optimizer step, enabling the weight update to be done more precisely
```
Comment (ravi-mosaicml) [inline docs/source/trainer/numerics.md:56]:
```suggestion
3. To increase numerical sensitivity, convert the model to single-precision and perform the optimizer step, enabling the weight update to be done more precisely.
```

File: docs/source/trainer/numerics.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+- Gradient underflow: Very small values can get zeroed out
+- Incorrect weight updates: Small gradient values can lead to no weight updates
+- Activation/Loss overflow: Very large values can roll over to representations of [Infinity](https://en.wikipedia.org/wiki/Infinity#Computing)(`Inf`)or [Not a Number](https://en.wikipedia.org/wiki/NaN) (`NaN`)
+
+The result is reduced model accuracy.
+
+The solution is to perform mixed precision training, where both single and half-precision formats are utilized strategically.  Mixed precision training is performed as follows:
+
+1. Compute the forward pass and loss in half-precision, *except* for computations that can cause activations/loss overflow, which are performed in single-precision (e.g., BatchNorm)
+2. Compute the backpropagation pass in half-precision
```
Comment (ravi-mosaicml) [inline docs/source/trainer/numerics.md:55]:
```suggestion
2. Compute the backpropagation pass in half-precision.
```

File: docs/source/trainer/numerics.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+- Gradient underflow: Very small values can get zeroed out
+- Incorrect weight updates: Small gradient values can lead to no weight updates
+- Activation/Loss overflow: Very large values can roll over to representations of [Infinity](https://en.wikipedia.org/wiki/Infinity#Computing)(`Inf`)or [Not a Number](https://en.wikipedia.org/wiki/NaN) (`NaN`)
+
+The result is reduced model accuracy.
+
+The solution is to perform mixed precision training, where both single and half-precision formats are utilized strategically.  Mixed precision training is performed as follows:
+
+1. Compute the forward pass and loss in half-precision, *except* for computations that can cause activations/loss overflow, which are performed in single-precision (e.g., BatchNorm)
```
Comment (ravi-mosaicml) [inline docs/source/trainer/numerics.md:54]:
```suggestion
1. Compute the forward pass and loss in half-precision, *except* for computations that can cause activations/loss overflow, which are performed in single-precision (e.g., BatchNorm).
```

File: docs/source/trainer/numerics.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+2. Compute the backpropagation pass in half-precision
+3. To increase numerical sensitivity, convert the model to single-precision and perform the optimizer step, enabling the weight update to be done more precisely
+4. Convert the model back to half-precision
+
+The procedure above mitigates the issues with overflow and imprecise weight updates. 
+
+#### Gradient Scaling
+
+In order to prevent gradient underflow, mixed precision training uses a technique called **Gradient Scaling** where the loss is multiplied by a scaling factor *before* backpropagation is performed (step 2 above). This trick scales the computed gradient values to a range where they cannot underflow.  Once the model has been converted to single-precision following the backward pass, the gradients are divided by the scaling factor to restore the gradient values to the original range before stepping the optimizer.
```
Comment (ravi-mosaicml) [inline docs/source/trainer/numerics.md:63]: Maybe elaborate how gradient scaling automatically tunes itself? It catches nan/inf errors and then adjusts the scaling factor for the next batch. So batches that had nan/infs are skipped.

File: docs/source/trainer/numerics.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+### AMP
+
+Composer supports automatic mixed precision (AMP) training using PyTorch’s {mod}`torch.cuda.amp` package. The Composer {class}`~.Trainer` performs all the heavy lifting and tensor conversions automatically; the user simply has to set `precision='amp'` when initializing the {class}`~.Trainer`.
+
+Another advantage of using Composer is support for optimizer closures when using mixed precision.  A closure is responsible for clearing the gradients and re-computing the loss, as required by some optimizers. The {class}`~.Trainer` supports closures when using `amp` if the optimizer object has the attribute {attr}`~.SAMOptimizer._step_supports_amp_closure` set to `True`.  Please see the [SAM Method Card](../method_cards/sam.md) for an implementation example on how closures are used.
```
Comment (ravi-mosaicml) [inline docs/source/trainer/numerics.md:69]: Do we have an API docs reference for `step_supports_amp_closure`? If not, we should somewhere. But for now, could you rephrase this as an example of how SAM is an example of how to use this flag?

File: docs/source/trainer/numerics.md [hunk 12]
```diff
@@ -0,0 +1,73 @@
+# 🔢 Numerics
+
+Composer supports several single and half-precision number formats including IEEE single-precision `fp32`, IEEE half-precision `fp16`, Google’s truncated `bf16` format and support for mixed precision training `amp`.  The following formats are supported per accelerator:
+
+| Format | CPU | GPU - A100 | GPU - V100 | GPU - T4 |
+| --- | --- | --- | --- | --- |
```
Comment (ravi-mosaicml) [inline docs/source/trainer/numerics.md:6]: Nit: can you format the table so the columns align?

**Review by ravi-mosaicml — COMMENTED** [review]
Overall looks good!

File: docs/source/trainer/numerics.md [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+| bf16 | ➖ | ✅ | ➖ | ➖ |
+| fp16 | ➖ | ✅ | ✅ | ✅ |
+| fp32 (default) | ✅ | ✅ | ✅ | ✅ |
+| amp | ➖ | ✅ | ✅ | ✅ |
+
+When using the {class}`~.Trainer`, the number format can selected by specifying the `precision` argument during initialization. In the example below, we are training on a `gpu` device and using the `fp16` format:
```
Comment (dskhudia) [inline docs/source/trainer/numerics.md:12]:
```suggestion
When using the {class}`~.Trainer`, the number format can selected by specifying the `precision` argument during initialization. In the example below, we are training on a `gpu` device using Automatic Mixed Precision (AMP). AMP is explained below:
```

The example shows automatic mixed-precision so the line " In the example below, we are training on a `gpu` device and using the `fp16` format" is incorrect.

**Comment by dskhudia** [comment]
@bandish-shah : Does it make sense to combine Mixed-precision and AMP section as they do talk about the same thing?

File: docs/source/trainer/numerics.md (hunk 3 above)
Comment (bandish-shah) [reply docs/source/trainer/numerics.md:44]: Handled by merging the AMP section with the Mixed Precision section.

File: docs/source/trainer/numerics.md (hunk 12 above)
Comment (bandish-shah) [reply docs/source/trainer/numerics.md:6]: I cleaned it up, won't be perfect bc the inline emoji's aren't same fixed fidth.

**Comment by bandish-shah** [comment]
> @bandish-shah : Does it make sense to combine Mixed-precision and AMP section as they do talk about the same thing?

Yes @ravi-mosaicml had the same question, combined in the latest iteration.

File: docs/source/trainer/numerics.md (hunk 11 above)
Comment (bandish-shah) [reply docs/source/trainer/numerics.md]: We don't, I added a cross-reference to the `SAMOptimizer` API Reference here for the time being.

**Review by dblalock — COMMENTED** [review]
Just local phrasing things

**Review by dskhudia — APPROVED** [review]
Feel free to merge once all comments from Davis are handled.

File: docs/source/trainer/numerics.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+The result can be reduced model accuracy or training divergence.
+
+The solution is to perform mixed precision training, where both single (`fp32`) and half-precision (`fp16`) formats are utilized strategically to avoid the issues above.  Composer supports Automatic Mixed Precision (AMP) training using PyTorch’s {mod}`torch.cuda.amp` package. The Composer {class}`~.Trainer` performs all the heavy lifting and tensor conversions automatically; the user simply has to set `precision='amp'` when initializing the {class}`~.Trainer`.
+
+Mixed precision training is performed as follows:
+
+1. Compute the forward pass and loss in half-precision, *except* for computations that can cause activations/loss overflow, which are performed in single-precision (e.g., BatchNorm).
+2. Compute the backpropagation pass in half-precision.
+3. To increase numerical sensitivity, convert the model to single-precision and perform the optimizer step, enabling the weight update to be done more precisely.
```
Comment (dblalock) [inline docs/source/trainer/numerics.md:54]:
```suggestion
3. Store the weights and perform the optimizer step in single precision, enabling the weight update to be done more precisely.
```

**Comment by dblalock** [comment]
^ the comments I added to threads started by ravi are marked as resolved, but they're actually separate suggestions and not resolved (sorry about that; TIL you need to start a new thread or this will happen)

**Comment by bandish-shah** [comment]
> ^ the comments I added to threads started by ravi are marked as resolved, but they're actually separate suggestions and not resolved (sorry about that; TIL you need to start a new thread or this will happen)

Think I got them all thanks!
